### PR TITLE
exec2: implement dynamic workload users taskrunner hook

### PR DESCRIPTION
**Pull request by shoenig**
This PR impelements a TR hook for allocating dynamic workload users from
a pool managed by the Nomad client. This adds a new task driver Capability,
DynamicWorkloadUsers - which a task driver must indicate in order to make
use of this feature.

The client config plumbing is coming in a followup PR - in the RFC we
realized having a client.users block would be nice to have, with some
additional unrelated options being moved from the deprecated client.options
config.

### CODE REVIEW

File: plugins/drivers/driver.go [hunk 1]
```diff
@@ -178,6 +178,12 @@ type Capabilities struct {
 	// DisableLogCollection indicates this driver has disabled log collection
 	// and the client should not start a logmon process.
 	DisableLogCollection bool
+
+	// DynamicWorkloadUsers indicates this driver is capable (but not required)
+	// of making use of UID/GID not backed by a user known to the operating system.
+	// The allocation of a unique, not-in-use UID/GID is managed by Nomad client
+	// ensuring no overlap.
+	DynamicWorkloadUsers bool
```
Comment (tgross) [inline plugins/drivers/driver.go:186]: I think this needs to get added to the `DriverCapabilities` protobuf message as well in the `plugins/drivers/proto/driver.proto`

File: helper/users/dynamic/users.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+	return UGID(i), err
+}
+
+// LookupUser will return the UID, GID, and home directory associated with the
+// given username. If username is of the form 'nomad-<id>' this indicates Nomad
+// has synthesized a dynamic workload user for the task and the UID/GID are the
+// <id> value.
+func LookupUser(username string) (int, int, string, error) {
+	// if we can succusfully parse username as an anonymous user, use that
```
Comment (tgross) [inline helper/users/dynamic/users.go:53]:
```suggestion
	// if we can successfully parse username as an anonymous user, use that
```

**Review by tgross — APPROVED** [review]
LGTM!

File: plugins/drivers/driver.go (hunk 1 above)
Comment (shoenig) [reply plugins/drivers/driver.go:186]: Yup will do in a followup; just getting the hook hooked up for now.

**Pull request merged**